File: MatlabCodes/Fig09_13.m
function BrownianBridge
clc;close all;

NoOfPaths  =6;
NoOfSteps  = 1000; 
T          = 2;
vt0        = 0.1;
vtN        = 0.2;
sigma1      =0.03;
sigma2      =0.01;

figure(1)
randn('seed',1)
[B,timeGrid] = GeneratePathsBrownianBridge(NoOfPaths,NoOfSteps,T,vt0,vtN,sigma1);
plot(timeGrid,B,'color',[0,0.45,0.75],'linewidth',1.5)
axis([0,T,min(min(B)),max(max(B))])
grid on
xlabel('time')
ylabel('v(t)')
a = axis;
hold on
plot([0,T],[vt0,vtN],'MarkerFaceColor',[1 0 0],'MarkerEdgeColor',[1 0 0],'Marker','o',...
    'LineStyle','none',...
    'Color',[1 0 0])

figure(2)
randn('seed',1)
[B,timeGrid] = GeneratePathsBrownianBridge(NoOfPaths,NoOfSteps,T,vt0,vtN,sigma2);
plot(timeGrid,B,'color',[0,0.45,0.75],'linewidth',1.5)
axis([0,T,min(min(B)),max(max(B))])
grid on
xlabel('time')
ylabel('v(t)')
axis(a)
hold on
plot([0,T],[vt0,vtN],'MarkerFaceColor',[1 0 0],'MarkerEdgeColor',[1 0 0],'Marker','o',...
    'LineStyle','none',...
    'Color',[1 0 0])

function [B,time] = GeneratePathsBrownianBridge(NoOfPaths,NoOfSteps,T,a,b,sigma)

% Brownian bridge process approximated with the Euler discretization

B=zeros(NoOfPaths,NoOfSteps);
B(:,1) = a;

% Random noise

Z=random('normal',0,1,[NoOfPaths,NoOfSteps]);
W=zeros([NoOfPaths,NoOfSteps]);

dt = T / NoOfSteps;
time = zeros([NoOfSteps+1,1]);
for i=1:NoOfSteps
    if NoOfPaths>1
        Z(:,i)   = (Z(:,i) - mean(Z(:,i))) / std(Z(:,i));
    end
    W(:,i+1)  = W(:,i) + sqrt(dt).*Z(:,i);
    B(:,i+1)  = B(:,i) +   (b-B(:,i))/(T -time(i))*dt + sigma*(W(:,i+1)-W(:,i));
    time(i+1) = time(i) + dt;
end

File: PythonCodes/Fig09_13.py
#%%
"""
Created on Thu Nov 27 2018
Brownian Bridge simulation for a given two points 
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt

def GenerateBrownianBridge(NoOfPaths,NoOfSteps,T,a,b,sigma):    
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    B = np.zeros([NoOfPaths, NoOfSteps+1])
    W = np.zeros([NoOfPaths, NoOfSteps+1])
    t = np.zeros([NoOfSteps+1])
    B[:,0]=a        
    dt = T / float(NoOfSteps)
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])
        W[:,i+1] = W[:,i] +  np.power(dt, 0.5)*Z[:,i]
        B[:,i+1] = B[:,i] +   (b-B[:,i])/(T -t[i])*dt + sigma*(W[:,i+1]-W[:,i])
        t[i+1] = t[i] +dt
        
    paths = {"time":t,"B":B}
    return paths

def mainCalculation():
    NoOfPaths = 6
    NoOfSteps = 2000
    T = 2.0
    V_t0 = 0.1
    V_tN = 0.2
    sigma = 0.03
    
    Paths = GenerateBrownianBridge(NoOfPaths,NoOfSteps,T,V_t0,V_tN,sigma)
    timeGrid = Paths["time"]
    B = Paths["B"]

    plt.figure(1)
    plt.plot(timeGrid, np.transpose(B),'-b')   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("B(t)")
    
                    
mainCalculation()
